presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.expressions.translator;
+
+import com.facebook.presto.spi.function.FunctionMetadata;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.google.common.collect.ImmutableMap;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.expressions.translator.TranslatedExpression.untranslated;
+import static java.util.Collections.emptyList;
+
+public class FunctionTranslator<T>
+{
+    private final Map<FunctionMetadata, MethodHandle> functionMapping;
+
+    public static <T> FunctionTranslator<T> buildFunctionTranslator(Set<Class<?>> translatorContainers)
+    {
+        ImmutableMap.Builder<FunctionMetadata, MethodHandle> functionMappingBuilder = new ImmutableMap.Builder<>();
+        translatorContainers.stream()
+                .map(TranslatorAnnotationParser::parseFunctionDefinitions)
+                .forEach(functionMappingBuilder::putAll);
+        return new FunctionTranslator<>(functionMappingBuilder.build());
+    }
+
+    public TranslatedExpression<T> translate(FunctionMetadata functionMetadata, RowExpression original, List<TranslatedExpression<T>> translatedArguments)
+    {
+        if (!isFunctionPresent(functionMetadata)) {
+            return untranslated(original);
```

Comment:
return:
```
   return new TranslatedExpression<>(Optional.empty(), original, translatedArguments);
```
?  the information of translatedArguments might be lost here.

Reply:
Nice catch, i guess we have no reason to throw away translated arguments